### PR TITLE
PIN-3666 Avoid revocations when possible

### PR DESCRIPTION
**Pull request by nttdata-rtorsoli**

### CODE REVIEW

File: tenants-certified-attributes-updater/src/main/scala/it/pagopa/interop/tenantscertifiedattributesupdater/util/Utils.scala [hunk 1]
```diff
@@ -96,7 +96,9 @@ object Utils {
         }
 
     val revocations: Map[PersistentExternalId, List[AttributeInfo]] =
-      fromTenant.toList
+      fromTenant
+        .filter(_._2.exists(_.revocationTimestamp.isEmpty))
```
Comment (TonioGela) [inline tenants-certified-attributes-updater/src/main/scala/it/pagopa/interop/tenantscertifiedattributesupdater/util/Utils.scala]: Can you name parameters? That's incredibly obscure.
Try something like

```scala
.filter {
 case (_, nameOfTheResource) => nameOfTheResource.exists(_.revocationTimestamp.isEmpty)
}
```

Comment (galales) [reply tenants-certified-attributes-updater/src/main/scala/it/pagopa/interop/tenantscertifiedattributesupdater/util/Utils.scala]: Agree, plus: isn't this solving the problem just partially?

For example:
- the tuple is (Tenant1, List(Attribute1, Attribute2))
- Attribute1 is already revoked
- Attribute2 is not revoked

The logic would still revoke both Attribute1 and Attribute2, right?

Comment (TonioGela) [reply tenants-certified-attributes-updater/src/main/scala/it/pagopa/interop/tenantscertifiedattributesupdater/util/Utils.scala]: That will then require a list unpack with a flatmap or so

Comment (nttdata-rtorsoli) [reply tenants-certified-attributes-updater/src/main/scala/it/pagopa/interop/tenantscertifiedattributesupdater/util/Utils.scala]: The code filter the list of attributes when revocationTimestamp is defined.

File: tenants-certified-attributes-updater/src/main/scala/it/pagopa/interop/tenantscertifiedattributesupdater/util/Utils.scala [hunk 2]
```diff
@@ -96,7 +96,11 @@ object Utils {
         }
 
     val revocations: Map[PersistentExternalId, List[AttributeInfo]] =
-      fromTenant.toList
+      fromTenant
+        .filter { case (_, attrs) =>
+          attrs.exists(_.revocationTimestamp.isEmpty)
```
Comment (galales) [inline tenants-certified-attributes-updater/src/main/scala/it/pagopa/interop/tenantscertifiedattributesupdater/util/Utils.scala]: [this point](https://github.com/pagopa/interop-be-jobs/pull/146#discussion_r1277311182) is still true.
If you prefer, attributes are unpacked at line 104, you can filter by revocationTimestamp after that

Comment (nttdata-rtorsoli) [reply tenants-certified-attributes-updater/src/main/scala/it/pagopa/interop/tenantscertifiedattributesupdater/util/Utils.scala]: I am not be able to understand why this filter at line 104 is more efficient than at line 99.

Comment (galales) [reply tenants-certified-attributes-updater/src/main/scala/it/pagopa/interop/tenantscertifiedattributesupdater/util/Utils.scala]: it's not about efficiency, but about correctness.
Your filter takes all tenants that contains at least one attribute without revocation timestamp.
This means that if a tenant contains both, the revocation is performed on both, not avoiding the revocations as we needed.

If you prefer we can see this point offline if it's still not clear

**Review by galales — APPROVED** [review]

**Review by galales — APPROVED** [review]

**Review by TonioGela — APPROVED** [review]

**Review by beetlecrunch — APPROVED** [review]

**Pull request merged**